cmake/modules/ImportedTargetHelpers.cmake

```diff
@@ -1,6 +1,177 @@
+	# this version check may be relatively expensive. Otherwise, log the error
+	# and inform the user.
+	if(_dvfs_run_result EQUAL "0")
+		set("${_version_out}" "${_run_output}" CACHE INTERNAL "Version of ${_target}")
+	else()
+		message(DEBUG "${_compile_output}")
+		message(
+			"${_target} was found but the version could not be determined automatically.\n"
+			"Set the cache variable `${_version_out}` to the version you have installed."
+		)
+	endif()
+endfunction()
+
+# Search for a package using config mode. If this fails to find the desired
+# target, use the specified fallbacks and add the target if they succeed. Set
+# the variables `prefix_LIBRARY`, `prefix_INCLUDE_DIRS`, and `prefix_VERSION`
+# if found for the caller to pass to `find_package_handle_standard_args`.
+#
+# Usage:
+# 	_find_package_config_mode_with_fallback(
+#		<package_name>                 # The package to search for with config mode
+#		<target_name>                  # The target to expect from config mode, or define if not found
+# 		LIBRARY_NAMES names...         # Possible library names to search for as a fallback
+# 		INCLUDE_NAMES names...         # Possible header names to search for as a fallback
+# 		[PKG_CONFIG <pkg_config_name>] # The pkg-config name to search for as a fallback
+# 		[LIBRARY_HINTS hints...]       # Locations to look for libraries
+# 		[INCLUDE_HINTS hints...]       # Locations to look for headers
+#		[DEPENDS dependencies...]      # Dependencies of the target - added to INTERFACE_LINK_LIBRARIES, and will fail if not found
+#		[PREFIX <prefix>]              # The prefix for result variables - defaults to the package name
+# 	)
+function(_find_package_config_mode_with_fallback _fpcmwf_PACKAGE_NAME _fpcmwf_TARGET_NAME)
+	# Parse remaining arguments
+	set(_options "")
+	set(_one_value_args "PKG_CONFIG" "PREFIX")
+	set(_multi_value_args "LIBRARY_NAMES" "LIBRARY_HINTS" "INCLUDE_NAMES" "INCLUDE_HINTS" "DEPENDS")
+	cmake_parse_arguments(PARSE_ARGV 2 _fpcmwf "${_options}" "${_one_value_args}" "${_multi_value_args}")
+
+	# Compute result variable names
+	if(NOT DEFINED _fpcmwf_PREFIX)
+		set(_fpcmwf_PREFIX "${_fpcmwf_PACKAGE_NAME}")
+	endif()
+	set(_version_var "${_fpcmwf_PREFIX}_VERSION")
+	set(_library_var "${_fpcmwf_PREFIX}_LIBRARY")
+	set(_include_var "${_fpcmwf_PREFIX}_INCLUDE_DIRS")
+
+	# Try config mode if possible
+	find_package("${_fpcmwf_PACKAGE_NAME}" CONFIG QUIET)
+
+	if(TARGET "${_fpcmwf_TARGET_NAME}")
+		# Extract package details from existing target
+		get_target_property("${_library_var}" "${_fpcmwf_TARGET_NAME}" LOCATION)
+		get_target_property("${_include_var}" "${_fpcmwf_TARGET_NAME}" INTERFACE_INCLUDE_DIRECTORIES)
+		if(DEFINED "${_fpcmwf_PACKAGE_NAME}_VERSION")
+			set("${_version_var}" "${${_fpcmwf_PACKAGE_NAME}_VERSION}")
+		endif()
+	else()
+		# Check whether the dependencies exist
+		foreach(_dependency IN LISTS _fpcmwf_DEPENDS)
+			if(NOT TARGET "${_dependency}")
+				return()
```

Comment:
I think we could use a status message here, explaining why an early return was made.

Reply:
I'd like this function to stay quiet for now - you'll notice it passes `QUIET` to `find_package` and `pkg_check_modules`. The find module that calls it should first search for all the dependencies, and if it fails to find them, that's where the error should be reported. Perhaps it could be useful to have a `QUIET` argument for this function too, and report any errors if it's not passed, but I'd rather do that if or when we need it.